crates/parking_spot/src/lib.rs

```diff
@@ -0,0 +1,477 @@
+//! Implements thread wait and notify primitives with `std::sync` primitives.
+//!
+//! This is a simplified version of the `parking_lot_core` crate.
+//!
+//! There are two main operations that can be performed:
+//!
+//! - *Parking* refers to suspending the thread while simultaneously enqueuing it
+//! on a queue keyed by some address.
+//! - *Unparking* refers to dequeuing a thread from a queue keyed by some address
+//! and resuming it.
+//!
+//! # Example
+//!
+//! ```rust
+//! use std::sync::atomic::{AtomicU64, Ordering};
+//! use std::sync::Arc;
+//! use std::thread;
+//! use parking_spot::ParkingSpot;
+//!
+//! let parking_spot: Arc<ParkingSpot> = Arc::new(ParkingSpot::default());
+//! let atomic: Arc<AtomicU64> = Arc::new(AtomicU64::new(0));
+//! let atomic_key = Arc::as_ptr(&atomic) as usize;
+//!
+//! let handle = {
+//!     let parking_spot = parking_spot.clone();
+//!     let atomic = atomic.clone();
+//!     thread::spawn(move || {
+//!         atomic.store(1, Ordering::Relaxed);
+//!         parking_spot.unpark_all(atomic_key);
+//!         parking_spot.park(atomic_key, || atomic.load(Ordering::SeqCst) == 1, None);
+//!         assert_eq!(atomic.load(Ordering::SeqCst), 2);
+//!     })
+//! };
+//!
+//! parking_spot.park(atomic_key, || atomic.load(Ordering::SeqCst) == 0, None);
+//! atomic.store(2, Ordering::Relaxed);
+//! parking_spot.unpark_all(atomic_key);
+//!
+//! handle.join().unwrap();
+//! ```
+#![deny(clippy::all)]
+#![deny(clippy::pedantic)]
+#![deny(missing_docs)]
+#![deny(unsafe_code)]
+
+use std::collections::BTreeMap;
+use std::sync::{Arc, Condvar, Mutex};
+use std::time::{Duration, Instant};
+
+/// Result of a park operation.
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+pub enum ParkResult {
+    /// Unparked by another thread.
+    Unparked,
+
+    /// The validation callback returned false.
+    Invalid,
+
+    /// The timeout expired.
+    TimedOut,
+}
+
+#[derive(Default, Debug)]
+struct Spot {
+    to_unpark: usize,
+    num_parked: usize,
```

Comment:
Could you add comments here as to what these fields are tracking? For example I believe `to_unpark` is required per the spec since it doesn't allow for spurious wakeups.

Reply:
done